File: idscp2/src/main/java/de/fhg/aisec/ids/idscp2/drivers/default_driver_impl/rat/tpm2d/TPM2dHelper.java
package de.fhg.aisec.ids.idscp2.drivers.default_driver_impl.rat.tpm2d;

import org.checkerframework.checker.nullness.qual.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.security.MessageDigest;
import java.security.SecureRandom;
import java.security.cert.Certificate;

public class TPM2dHelper {
  private static final Logger LOG = LoggerFactory.getLogger(TPM2dHelper.class);
  private static final SecureRandom sr = new SecureRandom();

  private TPM2dHelper() {}

  /**
   * Generate a crypto-secure random hex String of length numChars
   *
   * @param numBytes Desired String length
   * @return The generated crypto-secure random hex String
   */
  static byte[] generateNonce(int numBytes) {
    byte[] randBytes = new byte[numBytes];
    sr.nextBytes(randBytes);
    return randBytes;
  }

  /**
   * Calculate SHA-1 hash of (nonce|certificate).
   *
   * @param nonce The plain, initial nonce
   * @param certificate The certificate to hash-combine with the nonce
   * @return The new nonce, updated with the given certificate using SHA-1
   */
  static byte[] calculateHash(byte[] nonce, @Nullable Certificate certificate) {
    try {
      MessageDigest digest = MessageDigest.getInstance("SHA-1");
      digest.update(nonce);
      if (certificate != null) {
        digest.update(certificate.getEncoded());
      } else {
        if (LOG.isWarnEnabled()) {
          LOG.warn(
              "No client certificate available. Cannot bind nonce to public key to prevent masquerading attack. TLS misconfiguration!");
        }
      }
      return digest.digest();
    } catch (Exception e1) {
      LOG.error("Could not create hash of own nonce and local certificate", e1);
      return nonce;
    }
  }

  static class ByteArrayUtil {

    private static final String[] lookup = new String[256];

    static {
      for (int i = 0; i < lookup.length; ++i) {
        if (i < 16) {
          lookup[i] = "0" + Integer.toHexString(i);
        } else {
          lookup[i] = Integer.toHexString(i);
        }
      }
    }

    static String toPrintableHexString(byte[] bytes) {
      StringBuilder s = new StringBuilder();
      for (int i = 0; i < bytes.length; ++i) {
        if (i > 0 && i % 16 == 0) {
          s.append('\n');
        } else {
          s.append(' ');
        }
        s.append(lookup[bytes[i] & 0xff]);
      }
      return s.toString();
    }
  }

}

File: idscp2/src/main/kotlin/de/fhg/aisec/ids/idscp2/drivers/default_driver_impl/rat/tpm2d/TPM2dHelper.kt
package de.fhg.aisec.ids.idscp2.drivers.default_driver_impl.rat.tpm2d

import org.slf4j.LoggerFactory
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.cert.Certificate

object TPM2dHelper {
    private val LOG = LoggerFactory.getLogger(TPM2dHelper::class.java)
    private val sr = SecureRandom()

    /**
     * Generate a crypto-secure random hex String of length numChars
     *
     * @param numBytes Desired String length
     * @return The generated crypto-secure random hex String
     */
    fun generateNonce(numBytes: Int): ByteArray {
        val randBytes = ByteArray(numBytes)
        sr.nextBytes(randBytes)
        return randBytes
    }

    /**
     * Calculate SHA-1 hash of (nonce|certificate).
     *
     * @param nonce       The plain, initial nonce
     * @param certificate The certificate to hash-combine with the nonce
     * @return The new nonce, updated with the given certificate using SHA-1
     */
    fun calculateHash(nonce: ByteArray, certificate: Certificate?): ByteArray {
        return try {
            val digest = MessageDigest.getInstance("SHA-1")
            digest.update(nonce)
            if (certificate != null) {
                digest.update(certificate.encoded)
            } else {
                if (LOG.isWarnEnabled) {
                    LOG.warn(
                            "No client certificate available. Cannot bind nonce to public key to prevent masquerading attack. TLS misconfiguration!")
                }
            }
            digest.digest()
        } catch (e1: Exception) {
            LOG.error("Could not create hash of own nonce and local certificate", e1)
            nonce
        }
    }

    internal object ByteArrayUtil {
        private val lookup = arrayOfNulls<String>(256)
        fun toPrintableHexString(bytes: ByteArray): String {
            val s = StringBuilder()
            for (i in bytes.indices) {
                if (i > 0 && i % 16 == 0) {
                    s.append('\n')
                } else {
                    s.append(' ')
                }
                s.append(lookup[bytes[i].toInt() and 0xff])
            }
            return s.toString()
        }

        init {
            for (i in lookup.indices) {
                if (i < 16) {
                    lookup[i] = "0" + Integer.toHexString(i)
                } else {
                    lookup[i] = Integer.toHexString(i)
                }
            }
        }
    }
}